### ISSUE TITLE
MSL backend put_expression() uses 45K of the stack

### ISSUE THREAD

**Issue by kvark** [gfx-rs/naga#595]
Doing so would quickly get us into the nasty stack overflow territory.
[Function code](https://github.com/gfx-rs/naga/blob/dd2cd923315fd1cc36ed873f1d0e8751e2599144/src/back/msl/writer.rs#L201-L642).

**Comment by kvark** [comment]
Blocked by https://github.com/japaric/cargo-call-stack/issues/32 to investigate (on macOS at least).

**Comment by kvark** [comment]
Methodology:
```rust
    fn put_expression(..) {
        let x = 0u32;
        println!("{:p}", &x as *const _);
        ...
    }
```
I see this in debug:
> 0x7ffeef1d6cb4
0x7ffeef1cbc44
0x7ffeef1d6cb4
0x7ffeef1cbc44

So one debug frame is 0x7ffeef1d6cb4 - 0x7ffeef1cbc44 == 45168 bytes.

In release:
> 0x7ffee800ffe0
0x7ffee8010e80
0x7ffee8010b80
0x7ffee8010990

So one release frame is 0x7ffee8010e80 - 0x7ffee8010b80 == 768 bytes. That's much better!

**Comment by jrmuizel** [comment]
Here's an rough analysis of the stack frame using ddbug https://gist.github.com/jrmuizel/4fdfbcd866038aa8594bd1a2b6e1c9cf

I haven't tried to interpret it yet.

**Comment by kvark** [comment]
Thanks @jrmuizel !
So all those `write!` and `?` aren't free after all :)

The biggest offender is `enum Error` taking 96 bytes and repeating a whole bunch of stuff on the stack (because as you mentioned in the chat, in Debug the stack items are not coalesced properly).

Here is what we need:

(1) A robust mechanism to track the stack usage of functions. Something like:
```rust
#[check_stack_size(100)] // would fail at build time if exceeded
fn foo() {}
```

(2) Reduce the size of `Error` drastically. It comes down to this variant:
```rust
IncompatibleOperands {
        op: String,
        left: String,
        right: String,
    },
```

**Comment by jrmuizel** [comment]
In the stack_frame section the lines are "offset[size] name: type".

There's a lot of '<unknown>' which I'm guessing is used for storing values that aren't named. The rest of the space is mostly used by small types.

It doesn't look there's anything egregious, just a large accumulation of temporaries and so there's probably no easy way to fix this. Further evidence of this is the drastic drop in stack usage when using `--release`.

**Comment by jrmuizel** [comment]
Can you just `Box<>` to fix it's size issue?

**Comment by kvark** [comment]
That's one way to go about it, I suppose. Would probably require all the called functions to also return `Box<Error>`, in which case we'd be losing the ability to have `?` converting one error into another.

**Comment by jrmuizel** [comment]
I meant putting the IncompatibleOperands in a `Box`. i.e. try to keep Error just a discriminant and Box

**Comment by kvark** [comment]
Yes, that would be a good improvement!
It's a bit sad that `String`, having all the contents boxed anyway, is such big (24 bytes).

**Comment by jrmuizel** [comment]
https://crates.io/crates/thin_str?

**Comment by jrmuizel** [comment]
But really since it's just an error the extra boxing is unlikely to matter

**Comment by kvark** [comment]
It's still too much (around 20K), and we are overflowing...
Need to split it into sub-functions.
